DcWebxdcWidget/DcWebxdcWidget.swift

```diff
@@ -0,0 +1,177 @@
+            WebxdcApp(accountId: 0, chatId: 0, messageId: 3, image: UIImage(named: "webxdc"), title: "webxdc"),
+            WebxdcApp(accountId: 0, chatId: 0, messageId: 2, image: UIImage(named: "pixel"), title: "pixel"),
+            WebxdcApp(accountId: 0, chatId: 0, messageId: 4, image: UIImage(named: "checklist"), title: "checklist"),
+            WebxdcApp(accountId: 0, chatId: 0, messageId: 5, image: UIImage(named: "hello"), title: "hello"),
+            WebxdcApp(accountId: 0, chatId: 0, messageId: 7, image: UIImage(named: "webxdc"), title: "webxdc"),
+        ]
+
+        return UsedWebxdcEntry(date: Date(), apps: Array(apps.prefix(limit)))
+    }
+
+    func getSnapshot(in context: Context, completion: @escaping (UsedWebxdcEntry) -> Void) {
+        let entry = placeholder(in: context)
+        completion(entry)
+    }
+
+    func getTimeline(in context: Context, completion: @escaping (Timeline<Entry>) -> Void) {
+        /// ----------------------------------
+        /// **DO NOT** START dcAccounts.startIo() IN WIDGET
+        /// ----------------------------------
+        let dcAccounts = DcAccounts.shared
+        dcAccounts.openDatabase(writeable: false)
+        for accountId in dcAccounts.getAll() {
+            let dcContext = dcAccounts.get(id: accountId)
+            if dcContext.isOpen() == false {
+                do {
+                    let secret = try KeychainManager.getAccountSecret(accountID: accountId)
+                    _ = dcContext.open(passphrase: secret)
+                } catch {
+                    debugPrint("Couldn't open \(error.localizedDescription)")
+                }
+            }
+        }
+
+        let dcContext = dcAccounts.getSelected()
+        let chatId = 0
+        let ignore = Int32(0)
+
+        let limit: Int
+        switch context.family {
+        case .systemSmall: limit = 4
+        case .systemMedium: limit = 8
+        default: limit = 8
+        }
+
+        let messageIds: [Int] = Array(dcContext.getChatMedia(chatId: chatId, messageType: DC_MSG_WEBXDC, messageType2: ignore, messageType3: ignore).reversed().prefix(limit))
+
+        let apps = messageIds.compactMap {
+            dcContext.getMessage(id: $0)
+        }.compactMap { msg in
+            let name = msg.getWebxdcAppName()
+            let image = msg.getWebxdcPreviewImage()
+            let accountId = dcContext.id
+            let chatId = msg.chatId
+
+            return WebxdcApp(
+                accountId: accountId,
+                chatId: chatId,
+                messageId: msg.id,
+                image: image,
+                title: name
+            )
+        }
+
+        let currentDate = Date()
+        let entry = UsedWebxdcEntry(date: currentDate, apps: apps)
+        let nextDate = Calendar.current.date(byAdding: .minute, value: 15, to: currentDate)!
+
+        let timeline = Timeline(entries: [entry], policy: .after(nextDate))
+        completion(timeline)
```

Comment:
i am wondering, how other apps create "shortcut widgets" - are there other providers than "TimelineProvider"? just a general question for understanding, for the current implementation, things are fine

Reply:
From what I know, there's just the Timeline-provider. Basically you can't just update the widget itself [whenever you want](https://developer.apple.com/documentation/WidgetKit/Keeping-a-Widget-Up-To-Date), just like with background activity, you have a limited reload-budget and the system is in control to allocate your widget some time to reload here and there throughout the day.

That's why you build a "Timeline" with entries and give a "Please update then"-date. Does this answer your question?

Reply:
yes, so i learned it is always  a "timeline", thanks  for clarifying that, this was the part i was unsure about.

however,  as it is possible to let ["WidgetKit request a new timeline when something affects a widget’s current timeline"](https://developer.apple.com/documentation/WidgetKit/Keeping-a-Widget-Up-To-Date#Inform-WidgetKit-when-a-timeline-changes) it should be fine also for the case the user adds/removes widgets manually